vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+            //uint32_t random_pos = merand48(b.all->random_state)*(path_to_leaf.size()-1);
+            uint32_t random_pos = merand48(b.all->random_state)*(path_to_leaf.size()); //include leaf
+            uint32_t cn = path_to_leaf[random_pos];
+            
+            if (b.nodes[cn].internal != -1){ //if it's an internal node:' 
+                float objective = 0.f;
+                float prob_right = 0.5;
+                float coin = merand48(b.all->random_state) < prob_right ? 1.f : -1.f;
+                float weight = path_to_leaf.size()*1.f/(path_to_leaf.size() - 1.f);
+                if (coin == -1.f){ //go left
+                    float reward_left_subtree = return_reward_from_node(b,base, b.nodes[cn].left, ec, weight);
+                    objective = (1.-b.alpha)*log(b.nodes[cn].nl/b.nodes[cn].nr) + b.alpha*(-reward_left_subtree/(1.-prob_right))/2.;
+                }
+                else{ //go right:
+                    float reward_right_subtree= return_reward_from_node(b,base, b.nodes[cn].right, ec, weight);
+                    objective = (1.-b.alpha)*log(b.nodes[cn].nl/b.nodes[cn].nr) + b.alpha*(reward_right_subtree/prob_right)/2.;
+                }
+                
+                float ec_input_weight = ec.weight;
+                
+                MULTICLASS::label_t mc;
+                MULTILABEL::labels multilabels;
+                MULTILABEL::labels preds;
+                if (b.oas == false)
+                    mc = ec.l.multi;
+                else{
+                    multilabels = ec.l.multilabels;
+                    preds = ec.pred.multilabels;
+                }
+                
+                ec.weight = fabs(objective);
+                if (ec.weight >= 100.f) //crop the weight, otherwise sometimes cause NAN outputs.
+                    ec.weight = 100.f;
+                else if (ec.weight < .01f)
+                    ec.weight = 0.01f;
+                ec.l.simple = {objective < 0. ? -1.f : 1.f, 1.f, 0.};
+                base.learn(ec, b.nodes[cn].base_router);
+               
+                if (b.oas == false)
+                    ec.l.multi = mc;
+                else{
+                    ec.pred.multilabels = preds;
+                    ec.l.multilabels = multilabels;
+                }
+                ec.weight = ec_input_weight; //restore the original weight
+            }
+            else{ //if it's a leaf node:
+                float weight = 1.f; //float(path_to_leaf.size());
+                if (b.learn_at_leaf == true)
+                    learn_at_leaf_random(b, base, cn, ec, weight); //randomly sample one example, query reward, and update leaf learner
+
+                if (b.oas == true)
+                    train_one_against_some_at_leaf(b,base, cn, ec);
+            }
+        }
+        path_to_leaf.delete_v();
+    }
+
+    void insert_example_hal(memory_tree& b, single_learner& base, const uint32_t& ec_array_index, example& ec) 
```

Comment:
Can this be removed?

Reply:
removed, I changed the name to update_rew, i.e., a update using reward signals